site/_posts/2020-11-13-securing-flutter-oauth.md

```diff
@@ -0,0 +1,766 @@
+
+OK, that's fine. But if you run this code, you'll see it pops out to the system browser. Why not use an embedded webview? Well, because a webview is totally controlled by the native application displaying it, the [current mobile best practices](https://tools.ietf.org/html/rfc8252) for OAuth require you to use the system browser, which is not under that control. From section 8.12 of that document:
+
+> This best current practice requires that native apps MUST NOT use embedded user-agents to perform authorization requests and allows that authorization endpoints MAY take steps to detect and block authorization requests in embedded user-agents.  
+
+After that digression, here's `loginAction`:
+
+```dart
+Future<void> loginAction() async {
+  setState(() {
+    isBusy = true;
+    errorMessage = '';
+  });
+
+  try {
+    final AuthorizationTokenResponse result =
+        await appAuth.authorizeAndExchangeCode(
+      AuthorizationTokenRequest(
+        FUSION_AUTH_CLIENT_ID,
+        FUSION_AUTH_REDIRECT_URI,
+        issuer: 'https://$FUSION_AUTH_DOMAIN',
+        scopes: <String>['openid'],
+      ),
+    );
+    log('data: $result');
+    final Map<String, Object> profile =
+        await getUserDetails(result.accessToken);
+
+    debugPrint('response: $profile');
+    await secureStorage.write(
+          key: 'refresh_token', value: result.refreshToken);
+    var gravatar = Gravatar(profile['email']);
+    var url = gravatar.imageUrl(
+      size: 100,
+      defaultImage: GravatarImage.retro,
+      rating: GravatarRating.pg,
+      fileExtension: true,
+    );
+    setState(() {
+      isBusy = false;
+      isLoggedIn = true;
+      name = profile['given_name'];
+      picture = url;
+    });
+  } on Exception catch (e, s) {
+    debugPrint('login error: $e - stack: $s');
+
+    setState(() {
+      isBusy = false;
+      isLoggedIn = false;
+      errorMessage = e.toString();
+    });
+  }
+}
+// ...
+```
+
+`loginAction` called `getUserDetails`. What does this to? It takes the access token as a parameter and makes the HTTP GET request to the auth server. If successful, it will receive the user details in response. This is a standard OIDC call. This function is called inside the `initAction` function as well.
+
+```dart
+// ...
+Future<Map<String, Object>> getUserDetails(String accessToken) async {
+  const String url = 'https://$FUSION_AUTH_DOMAIN/oauth2/userinfo';
+  final http.Response response = await http.get(
+    url,
+    headers: <String, String>{'Authorization': 'Bearer $accessToken'},
+  );
+
+  if (response.statusCode == 200) {
+    return jsonDecode(response.body);
+  } else {
+    throw Exception('Failed to get user details');
+  }
+}
+// ...
+```
+
+When you can login, you want to be able to logout too. In `logoutAction` we remove the access token from secure storage and set the state back to initial values as shown in the code snippet below:
+
+```dart
```

Comment:
This is fine, but we should probably mention that this doesn't log the user out of FusionAuth and if they click the login button, they will automatically be logged back in to the Flutter app.

Reply:
OK. I asked if he could modify the code to have it call `/oauth2/logout` which we do in most other apps.